rfd/0197-better-async-state-management-with-tanstack-query.md

```diff
@@ -0,0 +1,266 @@
+
+## Why
+
+TanStack Query is a very popular open-source library for managing async state in React applications, with over 44k stars
+on GitHub.
+It is a powerful tool that provides easy to use hooks for fetching and caching data from the backend.
+
+## Current Setup
+
+As it stands, data fetching in Teleport is done using one of two different ways:
+
+### useAttempt
+
+This is a custom hook that is used to manage any state in Teleport. The attempt can have different statuses (none,
+loading, error, success).
+
+This results in a lot of boilerplate code in managing the data fetching & state storage, and error handling.
+
+#### Example query
+
+```tsx
+function fetchAccessList() {
+  scopedAttempt.setAttempt({ status: 'processing' });
+
+  accessManagementService
+    .fetchAccessList(accessListId)
+    .then(fetchedAccessList => {
+      const [modifiedAccessList, newPerms] = modifyAccessList(
+        fetchedAccessList,
+        accessLists,
+        ctx
+      );
+      setAccessList(modifiedAccessList);
+      setPerms(newPerms);
+      scopedAttempt.setAttempt({ status: 'success' });
+    })
+    .catch((e: Error) =>
+      scopedAttempt.setAttempt({ status: 'failed', statusText: e.message })
+    );
+}
+```
+
+This then has to be called with an effect that watches the access list ID:
+
+```tsx
+// The accessListId can change if a user clicks on a different
+// access list in the notification dropdown.
+useEffect(() => {
+  if (!accessList || accessList.id !== accessListId) {
+    fetchAccessList();
+  }
+}, [accessListId]);
+```
+
+#### Example mutation
+
+```tsx
+const { attempt, setAttempt } = useAttempt();
+const isDisabled = attempt.status === 'processing';
+
+function onOk() {
+  setAttempt({ status: 'processing' });
+  accessManagementService
+    .deleteAccessList(accessListId)
+    .then(() =>
+      history.push(cfg.getAccessListManagementRoute(), {
+        deletedAccessListId: accessListId,
+      })
+    )
+    .catch((e: Error) =>
+      setAttempt({ status: 'failed', statusText: e.message })
+    );
+}
+```
+
+### useAsync
+
+This is a newer custom hook that is used to manage async state in Teleport. It requires less boilerplate than
+`useAttempt`, as it handles running async operations and handling errors, but still requires a manual trigger in an
+effect,
+which causes data to not fetch until at least the component has rendered once.
+
+Whilst this works fine, it lacks an extended API that TanStack Query can provide.
+
+#### Example query
+
+```tsx
+const [fetchQrCodeAttempt, fetchQrCode] = useAsync((privilegeToken: string) =>
+  auth.createMfaRegistrationChallenge(privilegeToken, 'totp')
+);
+
+useEffect(() => {
+  fetchQrCode(privilegeToken);
+}, []);
+```
+
+#### Example mutation
+
+```tsx
+const [deleteRequestAttempt, runDeleteRequest] = useAsync(async () => {
+  await ctx.workflowService.deleteAccessRequest(props.requestId);
+  historyService.replace(cfg.getAccessRequestRoute());
+});
+```
+
+## TanStack Query
```

Comment:
no change needing the RFD of course because this is an implementation detail, but i'd like to see `useAttempt` and `useAsync` marked as deprecated and pointed to `useQuery` instead